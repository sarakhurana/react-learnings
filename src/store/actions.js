import axios from "axios";

export const INCREMENT = "INCREMENT";
export const DECREMENT = "DECREMENT";
export const ADD = "ADD";
export const SUBTRACT = "SUBTRACT";
export const CREATE_POST = "CREATE_POST";
export const EDIT_POST = "EDIT_POST";
export const GET_POSTS = "GET_POST";
export const DELETE_POST = "DELETE_POST";
export const GET_POSTS_FAILURE = "GET_POST_FAILURE";
export const FAVOURITE_POST = "FAVOURITE_POST";
export const UNFAVOURITE_POST = "UNFAVOURITE_POST";

export const getPosts = (posts) => ({
  type: GET_POSTS,
  payload: posts,
});

export const getPostFailure = () => ({
  type: GET_POSTS_FAILURE,
});

export const fetchPosts = async (dispatch) => {
  try {
    const response = await axios.get("http://localhost:3000/posts");
    const posts = response.data;
    dispatch(getPosts(posts));
  } catch (e) {
    dispatch(getPostFailure());
  }
};

export const createPost = (post) => ({
  type: CREATE_POST,
  payload: post,
});

export const deletePost = (index) => ({
  type: DELETE_POST,
  payload: index,
});

export const favouritePost = (index) => ({
  type: FAVOURITE_POST,
  payload: index,
});

export const editPost = (post) => ({
  type: EDIT_POST,
  payload: post
})