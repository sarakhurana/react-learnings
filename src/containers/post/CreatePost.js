import React, { useContext, useState } from "react";
import { Context } from "../../context/Context";
import { createPost } from "../../store/actions";
import { Formik, Form, Field } from "formik";
import uuid from "react-uuid";
import "./createPost.css";
import UploadImage from "../../components/posts/UploadImage";
import { isEmpty } from "lodash";
import { createPostService } from "../../services/CreatePostService";

const CreatePost = ({ isCreatePost, setCreatePost, getPosts }) => {
  const { dispatch } = useContext(Context);
  const [image, setImage] = useState(undefined);

  const handleCreatePost = async (postBody) => {
    const post = { postId: uuid(), body: postBody };
    const { result, error } = await createPostService(post);
    if (!error) setCreatePost(false);
    getPosts();
  };
  const handleDispatchClick = (postBody) => {
    dispatch(
      createPost({
        image: image,
        body: postBody,
        postId: uuid(),
        isFavourite: false,
        isEditMode: false,
      })
    );
    setCreatePost(false);
  };
  return (
    <>
      {isCreatePost ? (
        <div className="create-post-container">
          <Formik
            initialValues={{ post: "" }}
            onSubmit={(values) => {
              handleCreatePost(values.post);
              // handleDispatchClick(values.post);
            }}
          >
            {({ values }) => (
              <Form>
                <UploadImage image={image} setImage={setImage} />
                <Field
                  className="text-post-body"
                  data-testid="test-post-body"
                  type="text"
                  name="post"
                  placeholder="Enter your post"
                />
                <div className="btn-container">
                  <button
                    data-testid="test-create-post"
                    className="btn-post"
                    type="submit"
                    disabled={isEmpty(values.post)}
                  >
                    Post
                  </button>
                </div>
              </Form>
            )}
          </Formik>
        </div>
      ) : (
        <></>
      )}
    </>
  );
};

export default CreatePost;
