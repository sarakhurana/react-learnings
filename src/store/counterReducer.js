import * as actionTypes from './actions'

const initialState = {
  counter: 0,
};

const counterReducer = (state = initialState, action) => {
  const { type, payload } = action;
  switch (type) {
    case actionTypes.INCREMENT:
      return { ...state, counter: state.counter + 1 };
    case actionTypes.ADD:
      return { ...state, counter: state.counter + payload };
    case actionTypes.DECREMENT:
      return { ...state, counter: state.counter - 1 };
    case actionTypes.SUBTRACT:
      return { ...state, counter: state.counter - payload };
    default:
      return state;
  }
};

export default counterReducer;
