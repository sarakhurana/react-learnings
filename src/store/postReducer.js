import * as actionTypes from "./actions";

const initialState = {
  posts: [],
};

const postReducer = (state = initialState, action) => {
  const { type, payload } = action;
  switch (type) {
    case actionTypes.CREATE_POST: {    
      state.posts.unshift(payload);
      return state
    }
    case actionTypes.GET_POSTS:
      return { ...state, posts: state.posts.concat(payload) };
    case actionTypes.GET_POSTS_FAILURE:
      return state;
    case actionTypes.DELETE_POST: {
      const updatedArray = state.posts.filter(
        result => result.postId !== payload
      );
      return { ...state, posts: updatedArray };
    }
      case actionTypes.FAVOURITE_POST: {
        const posts=state.posts;
        const index=posts.findIndex(post=>post.postId===payload)
        const newStatus= !posts[index].isFavourite
        const updatedState=state.posts;
        updatedState[index]={...state.posts[index], isFavourite:newStatus}
        return {...state, posts:updatedState};
    }
    case actionTypes.EDIT_POST: {
      const posts=state.posts;
      const index=posts.findIndex(post=>post.postId===payload.postId)
      const updatedState=state.posts;
      updatedState[index]={ ...state.posts[index], title: payload.title, body: payload.body}
      return {...state, posts:updatedState};
  }
    default:
      return state;
  }
};

export default postReducer;
