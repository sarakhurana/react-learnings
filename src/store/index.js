import { combineReducers } from "redux";
import postReducer from "./postReducer";
import counterReducer from "./counterReducer";

const rootReducer = combineReducers({
  counter: counterReducer,
  post: postReducer,
});
export default rootReducer;
