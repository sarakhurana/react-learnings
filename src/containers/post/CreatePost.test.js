import { act, fireEvent, render, waitFor } from "@testing-library/react";
import { createPost, CREATE_POST } from "../../store/actions";
import CreatePost from "./CreatePost";
import configureMockStore from "redux-mock-store";
import axiosConfig from "../../api/axiosConfig";
import MockAdapter from "axios-mock-adapter";

const mockStore = configureMockStore();

describe("CreatePost", () => {
  const mockAxios = new MockAdapter(axiosConfig, {
    delayResponse: 200,
  });

  let store;
  beforeEach(() => {
    store = mockStore({ post: [] });
  });

  it("should render if isCreatePost is false", () => {
    const { asFragment } = render(
      <CreatePost isCreatePost={false} setCreatePost={jest.fn()} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render post form fields if iscreatePost is true ", () => {
    const { asFragment, getByTestId } = render(
      <CreatePost isCreatePost={true} setCreatePost={jest.fn()} />
    );

    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId("test-post-body")).toBeInTheDocument();
  });

  it("should change value in input field", () => {
    const { getByTestId } = render(
      <CreatePost isCreatePost={true} setCreatePost={jest.fn()} />
    );
    const postInput = getByTestId("test-post-body");

    fireEvent.change(postInput, { target: { value: "Post body" } });

    expect(postInput.value).toBe("Post body");
  });

  it("should call createPostService with give post onn button click", async () => {
    jest.spyOn(axiosConfig, "post").mockImplementation().mockResolvedValue({
      status: 200,
      data: true,
    });
   const getPostSpy = jest.fn();
    const { getByTestId } = render(
      <CreatePost
        isCreatePost={true}
        setCreatePost={jest.fn()}
        getPosts={getPostSpy}
      />
    );

    const postInput = getByTestId("test-post-body");
    fireEvent.change(postInput, { target: { value: "Post body" } });
    const button = getByTestId("test-create-post");
    fireEvent.click(button);

    await waitFor(() => expect(axiosConfig.post).toHaveBeenCalledTimes(1));
    await waitFor(() => expect(getPostSpy).toHaveBeenCalledTimes(1));
  });

  xit("should dispatch createPost", async () => {
    store.dispatch(
      createPost({
        body: "body",
        postId: 1,
        isFavourite: false,
        isEditMode: false,
      })
    );
    expect(store.getActions()).toEqual([
      {
        type: CREATE_POST,
        payload: {
          body: "body",
          postId: 1,
          isFavourite: false,
          isEditMode: false,
        },
      },
    ]);
  });
});
