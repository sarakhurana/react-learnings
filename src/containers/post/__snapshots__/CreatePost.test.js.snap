// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePost should render if isCreatePost is false 1`] = `<DocumentFragment />`;

exports[`CreatePost should render post form fields if iscreatePost is true  1`] = `
<DocumentFragment>
  <div
    class="create-post-container"
  >
    <form
      action="#"
    >
      <div
        class="App"
      >
        <div
          class="fileUploader "
        >
          <div
            class="fileContainer"
          >
            <img
              alt="Upload Icon"
              class="uploadIcon"
              src="UploadIcon.svg"
            />
            <p
              class=""
            >
              Maximum size file: 5MB
            </p>
            <div
              class="errorsContainer"
            />
            <button
              class="chooseFileButton "
              type="button"
            >
              Choose an image
            </button>
            <input
              accept="image/*"
              name=""
              type="file"
            />
            <div
              class="uploadPicturesWrapper"
            >
              <div
                style="position: relative; display: flex; align-items: center; justify-content: center; flex-wrap: wrap; width: 100%;"
              />
            </div>
          </div>
        </div>
      </div>
      <input
        class="text-post-body"
        data-testid="test-post-body"
        name="post"
        placeholder="Enter your post"
        type="text"
        value=""
      />
      <div
        class="btn-container"
      >
        <button
          class="btn-post"
          data-testid="test-create-post"
          disabled=""
          type="submit"
        >
          Post
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
